### PR TITLE
Avoid infinite loops with symlinks

### PR DESCRIPTION
**Pull request by ssbarnea**
Fixes: #332

## Summary of changes

Closes 

### Pull Request Checklist
- [ ] Changes have tests
- [x] News fragment added in [`changelog.d/`].
  _(See [documentation][PR docs] for details)_

[`changelog.d/`]: https://github.com/pypa/setuptools/tree/master/changelog.d
[PR docs]:
https://setuptools.readthedocs.io/en/latest/development/developer-guide.html#making-a-pull-request

### CODE REVIEW

File: setuptools/_distutils/filelist.py [hunk 1]
```diff
@@ -247,12 +247,22 @@ def _find_all_simple(path):
...
+        for dirname in dirnames:
+            st = os.stat(os.path.join(dirpath, dirname))
+            dirkey = st.st_dev, st.st_ino
+            if dirkey not in dirs:
+                dirs.add(dirkey)
+                scandirs.append(dirname)
+        dirnames[:] = scandirs
+        for f in filenames:
+            files.add(os.path.join(dirpath, f))
+    return files
```
Comment (webknjaz) [inline setuptools/_distutils/filelist.py:265]: This should be submitted to https://github.com/pypa/distutils first.

File: setuptools/__init__.py [hunk 2]
```diff
@@ -213,12 +213,22 @@ def _find_all_simple(path):
...
+        for dirname in dirnames:
+            st = os.stat(os.path.join(dirpath, dirname))
+            dirkey = st.st_dev, st.st_ino
+            if dirkey not in dirs:
+                dirs.add(dirkey)
+                scandirs.append(dirname)
+        dirnames[:] = scandirs
+        for f in filenames:
+            files.add(os.path.join(dirpath, f))
+    return files
```
Comment (webknjaz) [inline setuptools/__init__.py:231]: Could you add a regression test for this change?

File: changelog.d/2707.change.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Avoid symlink infinite loops -- by :user:`ssbarnea`
```
Comment (webknjaz) [inline changelog.d/2707.change.rst:1]: Does this sound better?
```suggestion
Avoid infinite symlink follow loops -- by :user:`ssbarnea`
```

**Comment by ssbarnea** [comment]
Lets wait until https://github.com/pypa/distutils/pull/44 is merged, and resume this one after.

**Comment by jaraco** [comment]
See #2714, where I'm integrating the changes based on the report/PR in distutils. If these changes are needed for Setuptools separately from distutils (i.e. while Setuptools still falls back to stdlib distutils), let's use the implementation from distutils.

**Pull request closed without merging**